### PR TITLE
Win32 char to unicode

### PR DESCRIPTION
**Pull request by mrexodia**
#7235

### CODE REVIEW

File: libr/debug/p/native/w32.c [hunk 1]
```diff
@@ -245,69 +243,68 @@ static int w32_dbg_init() {
 	w32dbg_SeDebugPrivilege ();
 	/* lookup function pointers for portability */
 	w32_DebugActiveProcessStop = (BOOL (WINAPI *)(DWORD))
-		GetProcAddress (GetModuleHandleA ("kernel32"),
+		GetProcAddress (GetModuleHandle (TEXT ("kernel32")),
```
Comment (mrexodia) [inline libr/debug/p/native/w32.c]: Use `GetModuleHandleW(L"kernel32")` here explicitly because UNICODE might not be defined and you always want the wide variant anyway...

Comment (mrexodia) [reply libr/debug/p/native/w32.c]: Also, just cache this because the module handle for kernel32 is being used here quite a lot!

File: libr/debug/p/native/w32.c [hunk 2]
```diff
@@ -422,12 +415,6 @@ static int debug_exception_event (DEBUG_EVENT *de) {
 			get_w32_excep_name(code),
 			(int)de->dwThreadId);
 		break;
-#if __MINGW64__ || _WIN64
-	/* STATUS_WX86_BREAKPOINT */
```
Comment (mrexodia) [inline libr/debug/p/native/w32.c]: Why was this removed?

File: libr/debug/p/debug_native.c [hunk 3]
```diff
@@ -1398,8 +1421,9 @@ static int r_debug_native_drx (RDebug *dbg, int n, ut64 addr, int sz, int rwx, i
  * we only handle the case for hardware breakpoints here. otherwise,
  * we let the caller handle the work.
  */
-static int r_debug_native_bp (RBreakpoint *bp, RBreakpointItem *b, bool set) {
+static int r_debug_native_bp (void *bp_, RBreakpointItem *b, bool set) {
```
Comment (mrexodia) [inline libr/debug/p/debug_native.c]: Why was that parameter changed?

**Comment by mrexodia** [comment]
ping @josediazfer

File: libr/debug/p/native/w32.c [hunk 4]
```diff
@@ -442,81 +429,70 @@ static int debug_exception_event (DEBUG_EVENT *de) {
...
-	filename = malloc(MAX_PATH+1);
-
-	/* Create a file mapping to get the file name. */
-	void* map = MapViewOfFile (handle_file_map, FILE_MAP_READ, 0, 0, 1);
-
-	if (!map) {
-		free (filename);
-		CloseHandle (handle_file_map);
-		return NULL;
+	filename = malloc ((MAX_PATH + 1) * sizeof (TCHAR));
```
Comment (mrexodia) [inline libr/debug/p/native/w32.c]: Use wchar_t explicitly, you never want ANSI anyway.

File: libr/debug/p/native/w32.c [hunk 5]
```diff
@@ -442,81 +429,70 @@ static int debug_exception_event (DEBUG_EVENT *de) {
...
+	TCHAR name[MAX_PATH];
+	TCHAR drive[3] =  TEXT (" :");
+	LPTSTR cur_drive = temp_buffer;
+	while (*cur_drive) {
 		/* Look up each device name */
-		*drive = *p;
-		if (QueryDosDeviceA (drive, name, MAX_PATH)) {
-			size_t name_length = strlen (name);
+		*drive = *cur_drive;
+		if (QueryDosDevice (drive, name, MAX_PATH)) {
```
Comment (mrexodia) [inline libr/debug/p/native/w32.c]: QueryDosDeviceW

File: libr/debug/p/debug_native.c (hunk 3 above)
Comment (josediazfer) [reply libr/debug/p/debug_native.c]: Removed  warning with  RBreakpointCallback type aka "typedef int (*RBreakpointCallback)(void *bp, RBreakpointItem *b, bool set);"

File: libr/debug/p/native/w32.c (hunk 2 above)
Comment (josediazfer) [reply libr/debug/p/native/w32.c]: Because case STATUS_WX86_BREAKPOINT will never happen here because w32_dbg_wait function already catch case STATUS_WX86_BREAKPOINT.

File: libr/debug/p/native/w32.c (hunk 1 above)
Comment (josediazfer) [reply libr/debug/p/native/w32.c]: UNICODE is already defined in file mk/platform.mk
You're right, not cached, w32 code need refactory a lot.

File: libr/debug/p/native/w32.c (hunk 4 above)
Comment (josediazfer) [reply libr/debug/p/native/w32.c]: Comply with both windows calling convention and data types. UNICODE is already defined in file mk/platform.mk, TCHAR is wchar_t.

File: libr/debug/p/native/w32.c (hunk 5 above)
Comment (josediazfer) [reply libr/debug/p/native/w32.c]: Comply with both windows calling convention and data types. UNICODE is already defined in file mk/platform.mk, TCHAR is wchar_t.

**Comment by josediazfer** [comment]
Tomorrow will merge in a new branch with radare master branch. I have nasty logs in this branch.

**Comment by josediazfer** [comment]
Tomorow i will make a new branch with this changes merged with radare master. I have a lot of nasty logs in this branch.

**Comment by mrexodia** [comment]
I don't think `TCHAR` should be used especially because it requires `UNICODE` to be defined, it gives the pretense that you can disable `UNICODE` and that it will then work in ANSI mode, but this is obviously not the case.

**Comment by josediazfer** [comment]
Use the "T" data types is for comply with windows calling convention (microsoft code and samples are using winapi without speficy unicode or ansi) but I believe that is more simple remove ANSI support and use "W" data types (aka TCHAR as WCHAR, LPTSTR as LPWSTR) and speficy unicode suffix in functions.
I'm working on it

**Comment by XVilka** [comment]
On the one hand - there may be need to run radare2 on the old systems - e.g. Windows 95. So what about not using a TCHAR, but some another build time switch? To build for Unicode Windows - modern Windows now, but also keep the ability to run radare2 in ANSI-only Windows systems.

**Comment by josediazfer** [comment]
I very much doubt if radare would compile/run in windows 95/98/me versions (versions that work only with ANSI ).  I think that radare would support XP or greatest versions.

**Comment by XVilka** [comment]
TinyC compiler can compile Windows 95 or OpenWatcom, and they can compile radare2... Not sure about running though.

**Comment by mrexodia** [comment]
Radare most definitely doesn't run on windows 98 or even 200 lol, it uses way too many APIs that are not at all present in those versions of Windows. And the use of "T" datatypes is completely unrelated to calling conventions...

**Comment by josediazfer** [comment]
Not using -DUNICODE we will have problems with some macro that define a text value.
Example: 
#define SE_DEBUG_NAME TEXT("SeDebugPrivilege")

**Comment by mrexodia** [comment]
You can just do `LookupPrivilegeValueW(NULL, L"SeDebugPrivilige", &luidDebug)` ?

**Comment by josediazfer** [comment]
I don´t like use immediate text constants instead of constants already defined in a header file (aka winnt.h).

**Comment by mrexodia** [comment]
All good, just make sure that everything also works when `UNICODE` is not defined then?

**Comment by josediazfer** [comment]
UNICODE will always be defined if we doesn't think support for windows old versión in radare.

**Comment by mrexodia** [comment]
I don't think it makes sense to discuss this any longer.
